aclk/agent_cloud_link.c

```diff
@@ -1443,75 +1481,174 @@ int aclk_update_alarm(RRDHOST *host, ALARM_ENTRY *ae)
+        error("Unsuported version of \"version\" message from cloud. Expected %d, Got %d", ACLK_VERSION_NEG_VERSION, cloud_to_agent->version);
+        return 1;
+    }
+    if(unlikely(!cloud_to_agent->min_version)) {
+        error("Min version missing or 0");
+        return 1;
+    }
+    if(unlikely(!cloud_to_agent->max_version)) {
+        error("Max version missing or 0");
+        return 1;
+    }
+    if(unlikely(cloud_to_agent->max_version < cloud_to_agent->max_version)) {
+        error("Max version (%d) must be >= than min version (%d)", cloud_to_agent->max_version, cloud_to_agent->min_version);
+        return 1;
+    }
+
+    if(unlikely(cloud_to_agent->min_version > ACLK_VERSION_MAX)) {
+        error("Agent too old for this cloud. Minimum version required by cloud %d. Maximum version supported by this agent %d.", cloud_to_agent->min_version, ACLK_VERSION_MAX);
+        aclk_kill_link = 1;
+        aclk_disable_runtime = 1;
+        return 1;
+    }
+    if(unlikely(cloud_to_agent->max_version < ACLK_VERSION_MIN)) {
+        error("Cloud version is too old for this agent. Maximum version supported by cloud %d. Minimum (oldest) version supported by this agent %d.", cloud_to_agent->max_version, ACLK_VERSION_MIN);
+        aclk_kill_link = 1;
+        return 1;
+    }
+
+    version = MIN(cloud_to_agent->max_version, ACLK_VERSION_MAX);
+
     ACLK_SHARED_STATE_LOCK;
-    if (unlikely(aclk_shared_state.agent_state == AGENT_INITIALIZING)) {
-        debug(D_ACLK, "Ignoring cloud request; agent not in stable state");
-        ACLK_SHARED_STATE_UNLOCK;
-        return 0;
+    if (unlikely(now_monotonic_usec() > aclk_shared_state.version_neg_wait_till)) {
+        errno = 0;
+        error("The \"version\" message came too late ignoring.");
+        goto err_cleanup;
+    }
+    if (unlikely(aclk_shared_state.version_neg)) {
+        errno = 0;
+        error("Version has already been set to %d", aclk_shared_state.version_neg);
+        goto err_cleanup;
     }
+    aclk_shared_state.version_neg = version;
     ACLK_SHARED_STATE_UNLOCK;
 
+    info("Choosing version %d of ACLK", version);
+
+    return 0;
+
+err_cleanup:
+    ACLK_SHARED_STATE_UNLOCK;
+    return 1;
+}
+
+struct {
+    char *name;
+    int(*fnc)(struct aclk_request *cloud_to_agent);
+} aclk_incoming_msg_types[] = {
+    { .name = "http",    .fnc = aclk_handle_cloud_request    },
+    { .name = "version", .fnc = aclk_handle_version_response },
+    { .name = NULL,      .fnc = NULL                         }
+};
+
+int aclk_handle_cloud_message(char *payload)
+{
+    struct aclk_request cloud_to_agent;
+    memset(&cloud_to_agent, 0, sizeof(struct aclk_request));
+
+    if (aclk_stats_enabled) {
+        ACLK_STATS_LOCK;
+        aclk_metrics_per_sample.cloud_req_recvd++;
+        ACLK_STATS_UNLOCK;
+    }
+
     if (unlikely(!payload)) {
         debug(D_ACLK, "ACLK incoming message is empty");
-        return 0;
+        goto err_cleanup;
```

Comment:
note: this works only because the cloud_to_agent is zero-ed out. The structures are valid only after json parsing. You could save some processing time by jumping to a new label right before ACLK stats recording.

Reply:
Well it happens only on an error and would only skip 4 null checks at the end so the impact is minimal
but I will improve it to be perfectionist :)